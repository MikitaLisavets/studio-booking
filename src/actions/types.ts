import * as Actions from './Actions';
import { ErrorRequest, User } from '../types/types';

export type SetLocaleAction = {
  type: typeof Actions.SET_LOCALE;
  payload: string;
}
export type LocaleAction = SetLocaleAction;

export interface SetErrorAction {
  type: typeof Actions.SET_ERROR;
  payload: ErrorRequest;
}
export interface ClearErrorAction {
  type: typeof Actions.CLEAR_ERROR;
}
export type ErrorActionType = SetErrorAction | ClearErrorAction;

export interface SetUserAction {
  type: typeof Actions.SET_USER;
  payload: User;
}
export interface ClearUserAction {
  type: typeof Actions.CLEAR_USER;
}
export type UserActionType = SetUserAction | ClearUserAction;