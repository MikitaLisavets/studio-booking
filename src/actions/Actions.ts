import * as types from './types';
import { AnyAction } from 'redux';
import { AppState, ErrorRequest, User } from '../types/types';
import * as Rest from '../utils/rest';
import { ThunkAction } from 'redux-thunk';
import { MAIN_ROUTE } from '../constants/navigation';
import { history } from '../appRoutes';

export const SET_LOCALE = 'SET_LOCALE';
export function setLocale(locale: string): types.SetLocaleAction {
  return {
    type: SET_LOCALE,
    payload: locale
  };
}

export const SET_ERROR = 'SET_ERROR';
export function setError(error: ErrorRequest): types.SetErrorAction {
  return {
    type: SET_ERROR,
    payload: error
  };
}

export const CLEAR_ERROR = 'CLEAR_ERROR';
export function clearError(): types.ClearErrorAction {
  return {
    type: CLEAR_ERROR
  };
}

export const SET_USER = 'SET_USER';
export function setUser(user: User): types.SetUserAction {
  return {
    type: SET_USER,
    payload: user
  };
}

export const CLEAR_USER = 'CLEAR_USER';
export function clearUser(): types.ClearUserAction {
  return {
    type: CLEAR_USER
  };
}

export const SET_IS_CODE_CONFIRM_REQUIRED = 'SET_IS_CODE_CONFIRM_REQUIRED';
export function setIsCodeConfirmRequired(payload: boolean): types.IsCodeConfirmRequiredAction {
  return {
    type: SET_IS_CODE_CONFIRM_REQUIRED,
    payload
  };
}

export function initApp(): ThunkAction<void, AppState, never, AnyAction> {
  return (dispatch): void => {
    Rest.rest.setErrorHandler(error => { dispatch(setError(error)); });
    Rest.updateSession(() => { dispatch(clearUser()); })
      .then((response) => {
        if (!response) return;
        dispatch(setUser(response.user));
      });
  };
}

export function login({ email, password }: Rest.LoginRequest): ThunkAction<void, AppState, never, AnyAction> {
  return (dispatch): void => {
    Rest.login({ email, password })
      .then((response) => {
        if (!response) return;
        dispatch(setUser(response.user));
        history.push(MAIN_ROUTE);
      });
  };
}

export function logout(): ThunkAction<void, AppState, never, AnyAction> {
  return (dispatch): void => {
    Rest.logout().then(() => {
      dispatch(clearUser());
      history.push(MAIN_ROUTE);
    });
  };
}

export function signUp({ email, phoneNumber, password }: Rest.SignUpRequest): ThunkAction<void, AppState, never, AnyAction> {
  return (dispatch): void => {
    Rest.signUp({ email, phoneNumber, password })
      .then((response) => {
        if (response) dispatch(setIsCodeConfirmRequired(!response.UserConfirmed));
      });
  };
}

export function confirmSignUp({ confirmationCode, email, password }: Rest.ConfirmSignUpRequest): ThunkAction<void, AppState, never, AnyAction> {
  return (dispatch): void => {
    Rest.confirmSignUp({ confirmationCode, email, password })
      .then((response) => {
        if (!response) return;
        dispatch(setUser(response.user));
        dispatch(setIsCodeConfirmRequired(false));
        history.push(MAIN_ROUTE);
      });
  };
}