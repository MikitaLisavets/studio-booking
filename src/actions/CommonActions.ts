import { Action, AnyAction } from 'redux';
import { AppState, ErrorRequest } from '../types';
import { rest } from '../utils/rest';
import { ThunkAction } from 'redux-thunk';

export const SET_LOCALE = 'SET_LOCALE';

interface SetLocaleAction {
  type: typeof SET_LOCALE;
  payload: string;
}

export type LocaleAction = SetLocaleAction;

export function setLocale(locale: string): LocaleAction {
  return {
    type: SET_LOCALE,
    payload: locale
  };
}

export const SET_ERROR = 'SET_ERROR';
export const CLEAR_ERROR = 'CLEAR_ERROR';

interface SetErrorAction {
  type: typeof SET_ERROR;
  payload: ErrorRequest;
}

interface ClearErrorAction {
  type: typeof CLEAR_ERROR;
}
export type ErrorActionType = SetErrorAction | ClearErrorAction;

export function setError(error: ErrorRequest): ErrorActionType {
  return {
    type: SET_ERROR,
    payload: error
  };
}
export function clearError(): ErrorActionType {
  return {
    type: CLEAR_ERROR
  };
}

export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
export function increment(): Action {
  return {
    type: INCREMENT_COUNTER,
  };
}

export const DECREMENT_COUNTER = 'DECREMENT_COUNTER';
export function decrement(): Action {
  return {
    type: DECREMENT_COUNTER,
  };
}

export function initApp(): ThunkAction<void, AppState, never, AnyAction> {
  return (dispatch): void => {
    rest.setErrorHandler(error => { dispatch(setError(error)); });
  };
}