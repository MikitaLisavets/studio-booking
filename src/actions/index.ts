import { Action } from 'redux';
import { Error } from '../types';

export const SET_LOCALE = 'SET_LOCALE';

interface SetLocaleAction {
  type: typeof SET_LOCALE;
  payload: string;
}

export type LocaleAction = SetLocaleAction;

export function setLocale(locale: string): LocaleAction {
  return {
    type: SET_LOCALE,
    payload: locale
  };
}

export const SET_ERROR = 'SET_ERROR';
export const CLEAR_ERROR = 'CLEAR_ERROR';

interface SetErrorAction {
  type: typeof SET_ERROR;
  payload: Error;
}

interface ClearErrorAction {
  type: typeof CLEAR_ERROR;
}
export type ErrorActionType = SetErrorAction | ClearErrorAction;

export function setError(error: Error): ErrorActionType {
  return {
    type: SET_ERROR,
    payload: error
  };
}
export function clearError(): ErrorActionType {
  return {
    type: CLEAR_ERROR
  };
}

export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
export function increment(): Action {
  return {
    type: INCREMENT_COUNTER,
  };
}

export const DECREMENT_COUNTER = 'DECREMENT_COUNTER';
export function decrement(): Action {
  return {
    type: DECREMENT_COUNTER,
  };
}