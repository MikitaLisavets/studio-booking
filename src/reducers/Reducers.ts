import * as Actions from '../actions/Actions';
import { AppState, ErrorRequest, User } from '../types/Types';
import { combineReducers, Action } from 'redux';

export const initialState: AppState = {
  user: null,
  error: null,
  locale: 'en',
  counter: 0
};

export function user(state = initialState.user, action: Actions.UserActionType): User | null {
  switch (action.type) {
  case Actions.SET_USER:
    return action.payload;
  case Actions.CLEAR_USER:
    return null;
  default:
    return state;
  }
}

export function error(state = initialState.error, action: Actions.ErrorActionType): ErrorRequest | null {
  switch (action.type) {
  case Actions.SET_ERROR:
    return !action.payload.statusCode
      ? { statusCode: 500, code: 'SystemException', message: action.payload.message}
      : action.payload;
  case Actions.CLEAR_ERROR:
    return null;
  default:
    return state;
  }
}

export function locale(state = initialState.locale, action: Actions.LocaleAction): string {
  switch (action.type) {
  case Actions.SET_LOCALE:
    return action.payload;
  default:
    return state;
  }
}

export function counter(state = initialState.counter, action: Action): number {
  switch (action.type) {
  case Actions.INCREMENT_COUNTER:
    return state + 1;
  case Actions.DECREMENT_COUNTER:
    return state - 1;
  default:
    return state;
  }
}

export default combineReducers({
  user,
  error,
  locale,
  counter
});