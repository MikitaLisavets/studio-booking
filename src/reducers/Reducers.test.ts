import * as Actions from '../actions/Actions';
import * as Reducers from './Reducers';

describe('Reducers', () => {
  // describe('initialState', () => {
  //   it('returns initial state snapshot', () => {
  //     expect(Reducers.initialState).toMatchSnapshot();
  //   });
  // });

  describe('locale', () => {
    const action: Actions.LocaleAction = {
      type:    Actions.SET_LOCALE,
      payload: 'ru'
    };

    it('returns updated locale on SET_LOCALE action', () => {
      expect(Reducers.locale('en', action)).toEqual('ru');
    });
  });

  describe('counter', () => {
    it('returns incremented counter on INCREMENT_COUNTER action', () => {
      expect(Reducers.counter(0, { type: Actions.INCREMENT_COUNTER })).toEqual(1);
    });

    it('returns decremented counter on DECREMENT_COUNTER action', () => {
      expect(Reducers.counter(0, { type: Actions.DECREMENT_COUNTER })).toEqual(-1);
    });
  });
});