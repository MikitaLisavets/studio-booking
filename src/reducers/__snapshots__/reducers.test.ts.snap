// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducers initialState returns initial state snapshot 1`] = `
Object {
  "counter": 0,
  "error": null,
  "locale": "en",
}
`;
